Added: Set 1 (#8)
prathamesh-sonpatki/o11y-wiki: README.md

## C

### Cardinality

Cardinality refers to the number of unique values in a set of data. In the context of monitoring systems, high cardinality can negatively impact performance, so it's important to manage the cardinality of metrics and labels

### Counter

As the name says, counters are used to monitor how frequently an event occurs within a program.
They help monitor metrics that increase monotonically and are exposed as time series.
For example, total requests to an API endpoint is a counter metric, it keeps increasing over time.

### CI

> CI (Continuous Integration) is a practice of integrating code changes into a shared repository frequently. CI helps detect and resolve conflicts early and ensures that new code integrates smoothly with the existing codebase.

Reviewer: ```suggestion
CI (Continuous Integration) integrates code changes into a common branch of a shared repository and runs automated tests against it every time the changes are merged, or a commit is made. This allows for detecting issues in the code faster because it is tested continuously, as the name says.
```

## C

### Cardinality

Cardinality refers to the number of unique attributes of a metric. Eg. for a metric which tracks
no. of requests to an HTTP service, the attributes such as user's email can explode the unique 
combinations of the metric because each user's email will result into a unique [time series](#time-series).
High cardinality can result into latent dashboards because it causes querying of data slow due to higher
computation and memory overhead.

### Counter

As the name says, counters are used to monitor how frequently an event occurs within a program.
They help monitor metrics that increase monotonically and are exposed as time series.
For example, total requests to an API endpoint is a counter metric, it keeps increasing over time.

### CI

CI (Continuous Integration) integrates code changes into a common branch of a shared repository and runs automated tests against it every time the changes are merged, or a commit is made. This allows for detecting issues in the code faster because it is tested continuously, as the name says.

### CD

> CD (Continuous Delivery) is a practice of delivering code changes to production frequently and reliably. CD aims to reduce the time between code changes and deployment, and to ensure that new code changes don't break existing functionality.

Reviewer: ```suggestion
CD (Continuous Delivery) is the practice of delivering code changes to production frequently and reliably. An example would be using GitHub Actions to run tests via CI, prepare a docker image for the app once the tests pass, and then trigger the deployment via CD.
```

as logs, metrics, traces, events. The single storage allows for far more
corelation than different data being stored at different places.

### DevOps

> DevOps is a set of practices that combines software development and operations to shorten the development life cycle and improve reliability. DevOps teams use automation and collaboration to deploy software quickly and reliably.

Reviewer: ```suggestion
DevOps is a set of practices that combines software development and operations to shorten the development life cycle and improve reliability. The idea behind the DevOps movement is to reduce silos between different teams and make software delivery faster and more reliable via automation and collaboration.
```

### Downsampling

> Downsampling is the process of reducing the amount of data points in a time series to make it more manageable, without losing too much detail, and can help to save storage and processing resources.

Reviewer: ```suggestion
Downsampling is the process of reducing the amount of data points in a time series to make it more manageable without losing the essence and can help reduce the cost of storage and performance of querying. 
```

## E

### Event

> An event is a significant occurrence within a system that may require attention. Events can be used to trigger alerts or notify operators of potential issues.

Reviewer: ```suggestion
Events are hard to define because everything is an event, but let's give it a try 😉. An event is primarily a change event that can mean a pod restart, deployment, or configuration flag change. Events are important because they affect the system's state externally and can help correlate incidents with metrics, traces, and logs.
```

## I

### Incident

> An incident refers to an unexpected event that causes a system or service to fail or degrade. Examples of incidents can include server crashes, network outages, data center failures, or security breaches.

Reviewer: ```suggestion
An incident is an unexpected event that causes a system or service to fail or degrade. Examples of incidents can include service degradation, downtimes, server crashes, network outages, data center failures, or security breaches.
```

## J

## K

### Kubernetes

Kubernetes is a popular open-source platform for managing containerized workloads and services. Kubernetes provides features for deploying, scaling, and managing containerized applications.

### KPIs

> KPIs (Key Performance Indicator) are specific metrics used to track progress towards a specific goal or objective. In the context of SRE, KPIs can be used to measure the performance and reliability of a system.

Reviewer: ```suggestion
KPIs (Key Performance Indicators) track progress toward a specific goal or objective. In the context of o11y, KPIs can be used to measure the performance and reliability of a system, such as response times, failure rates, or peak load.
```

## L

### Log

> Logs are an record of activities performed by a system or service over a set period of time. Logs help SRE teams to track behaviours of a system. 

Reviewer: ```suggestion
Logs record activities a system, service, or tool performs over a set time. They are easy to adopt but run into standardization challenges because different tools and languages can have different logging patterns. Logs can grow quickly but are extremely important for debugging a problem once identified.
```

## M

### Metrics

> Metrics are quantifiable measurement about the health of a system. They serve as KPI(Key Performance Indicators) about the health of a system or service, Metrics should be easy to understand and makes SRE teams more effective to track performance of their systems.

Reviewer: ```suggestion
Metrics are quantifiable measurements of the health of a system. They are the fastest and cheapest way to understand the health of the system. They are aggregated data essentially giving a bird's eye view of the system performance. An example of a metric is the throughput of an API or latency.
```

### Metric Life Cycle Management

> Metric Life Cycle Management is the process of defining, collecting, analyzing, and acting on metrics. It's important to manage the metric life cycle to ensure that metrics are useful and provide insights into the system.

Reviewer: ```suggestion
Metric Life Cycle Management defines, collects, analyzes, and acts on metrics. This essentially involves creating a pipeline of metrics that moves from ingestion, storage, query, and alerting stages. You can drop unused data, perform streaming aggregates, and provide control over how and who can query specific data and define alerting.
```

### m3DB

> [m3DB](https://m3db.io/) is an open-source distributed time series database that is optimized for high cardinality and high throughput. m3DB is commonly used for storing and querying metrics in large-scale monitoring systems.

Reviewer: ```suggestion
[m3DB](https://m3db.io/) is an open-source distributed time series database that is optimized for high cardinality and high throughput. m3DB is commonly used for storing and querying metrics in large-scale monitoring systems. It was originated at Uber.
```

[Prometheus](https://prometheus.io/) is an open-source systems monitoring and
alerting toolkit originally built at SoundCloud.

### PromQL

> PromQL is a query language used to retrieve and analyze metrics from Prometheus, a popular open-source monitoring system.

Reviewer: ```suggestion
PromQL is a query language that retrieves and analyzes metrics from Prometheus, a popular open-source monitoring system. It is like SQL but for time series data. A lot of other time series databases, such as [VictoriaMetrics](https://victoriametrics.com/), and [Levitate](https://last9.io/products/levitate/) are also PromQL compatible.
```

## S

### Span

> A span is a single unit of work in a distributed system, often represented by a trace. Spans can be used to identify the performance of individual components in a system.

Reviewer: ```suggestion
A span is a single unit of work in a distributed system and a primary building block of distributed tracing. A trace represents a user request or a transaction in distributed tracing. Traces are broken down into multiple spans. For example, a function call to authenticate a user during a request can be represented by a span.
```

### Serverless

> Serverless is a model where a cloud provider manages the infrastructure and automatically scales resources, allowing developers to focus on code. This model is becoming increasingly popular for building and deploying applications.

Reviewer: ```suggestion
Serverless is a deployment model where a cloud provider manages the infrastructure and automatically scales resources, allowing developers to not worry about managing servers.
```

### SRE

SRE (Site Reliability Engineering) is a discipline focused on improving and maintaining the reliability of systems. SREs use a data-driven approach to manage the performance and availability of systems.

### Service Discovery

> Service discovery is the process of automatically identifying and locating available services in a distributed system. Service discovery is important for managing the complexity of distributed systems.

Reviewer: ```suggestion
Service Discovery is automatically identifying and locating available services in a distributed system. Service Discovery is essential for managing the complexity of distributed systems because of the dynamic and ephemeral nature of the systems.
```

### Samples

> Samples are individual data points in a time series. Metrics are collected as samples at regular intervals and are used to analyze the behavior of a system over time.

Reviewer: ```suggestion
Samples are individual data points in a time series. Metrics are collected as samples at regular intervals and are used to analyze the behavior of a system over time. A lot of time series databases use samples as billing metric.
```

data against successive equally spaced points in time, organizations can
understand the underlying causes of trends or systemic patterns over time.

### TSDB

> TSDB stands for time-series database. TSDBs are databases optimized for timeestamp data or time-series data, which in the context of observability contains metrics about systems which are timestamped with metrics about a system or service's health.

Reviewer: ```suggestion
TSDB or Time Series Databases are software systems are built for storing and retrieving time series data. Time series data are measurements or events that are tracked, monitored, downsampled, and aggregated over time.
```

### Trace

> A trace is a detailed record of the path a request takes through a system. Traces can be used to understand the performance of a system and identify bottlenecks.

Reviewer: ```suggestion
A trace is the complete journey of a request or workflow as it moves from one part of the system to another. It is achieved by adding a common trace id as the request/action flows through all the hops.
```
